core/src/main/kotlin/com/justai/jaicf/api/routing/BotRoutingEngine.kt

```diff
@@ -0,0 +1,198 @@
+ *
+ *
+ * Scenario usage example:
+ * ```
+ * private val main = Scenario {
+ *  fallback {
+ *      if (request.input == "sc1") routing.route("sc1")
+ *      else if (request.input == "sc2") routing.route("sc2")
+ *      else reactions.say("Hello from MAIN bot!")
+ *  }
+ * }
+ *
+ * private val sc1 = Scenario {
+ *  fallback {
+ *      reactions.say("Hello from SC1")
+ *  }
+ * }
+ *
+ * private val sc2 = Scenario {
+ *  fallback {
+ *      reactions.say("Hello from SC1")
+ *  }
+ * }
+ * private val router = BotRoutingEngine(
+ *  main = "default-router" to BotEngine(main),
+ *  routables = mapOf("sc1" to BotEngine(sc1), "sc2" to BotEngine(sc2))
+ *  )
+ * ```
+ *
+ * See more usage examples at examples/multilingual-bot.
+ *
+ * @param main a [BotEngine] implementation with name, this name will be used as this router's name
+ * @param routables a map of engines with names to perform route from and to.][]
+ *
+ * @see BotRoutingApi scenario runtime api for routing.
+ * */
+class BotRoutingEngine(
+    main: Pair<String, BotEngine>,
+    routables: Map<String, BotEngine>,
+) : BotEngine(MOCK_SCENARIO), WithLogger {
+
+    var routerName = main.first
+    val routerDefaultEngine = main.second
+    internal val routables: MutableMap<String, BotEngine> = routables.toMutableMap()
+    internal val children: MutableList<BotRoutingEngine> = mutableListOf()
+    internal var parent: BotRoutingEngine? = null
+
+    init {
+        if (routables.isEmpty()) {
+            error("Collection of routable engines must not be empty")
+        }
+        if (routables.values.distinctBy { it.defaultContextManager }.size > 1) {
+            error("Collections of routable botEngines must have single shared context manager instance")
+        }
+        if (main.second is BotRoutingEngine) {
+            error("Main bot cannot be an instance of BotRoutingEngine")
+        }
+        // TODO: Recursive validation of children routing engines to prevent naming clash
+
+        routables.values.forEach { engine ->
+            if (engine is BotRoutingEngine) {
+                children.add(engine)
+                engine.parent = this
+            }
+        }
+        this.routables[routerName] = routerDefaultEngine
+    }
+
+    private fun processWithRouting(
+        request: BotRequest,
+        reactions: Reactions,
+        requestContext: RequestContext,
+        contextManager: BotContextManager?,
+        botContext0: BotContext,
+        routeRequest: RoutingRequest,
+        executionContext0: ExecutionContext? = null,
+    ) {
+
+        val engineName = routeRequest.toEngine
+        val lastRouter = routeRequest.fromRouter ?: routerName
+        logger.info("Process route request: ${request.input} with routing for engine: $engineName and router: $lastRouter")
+
+        var router = getRouterRecursive(lastRouter) ?: this
+        if (engineName in router.childrenNames) {
+            router = getChildDescriptor(engineName) ?: router
```

Comment:
if-condition implies getChildDescriptor != null, the same about else-if-condition
So we can simplify it like that, for example:
```kotlin
if (engineName in router.childrenNames) {
    router = children.first { it.routerName == engineName }
} else if (engineName == router.parent?.routerName) {
    router = router.parent!!
}
```

and get rid of get*Descriptor() functions

Reply:
UPD: above is true for `router = router.getChildDescriptor(engineName)`. Looks like a mistake in the code, doesn't it?